### ISSUE TITLE
D7 upgrades: Crash due to different column name in url_alias table - `language` vs `langcode` (in specific MySQL/MariaDB versions)

### ISSUE THREAD

**Issue by irinaz** [backdrop/backdrop-issues#5496]
## Description of the bug

When I am setting D7 to backdrop on Pantheon following steps described in https://fibonacciwebstudio.com/nota-bene/d7-backdropcms-migration-pantheon-webuisftp-option I get the following error 

`SQLSTATE[HY000]: General error: 1034 Incorrect key file for table 'url_alias';`
This is similar to the problem described in a forum post  https://forum.backdropcms.org/forum/d7-conversion-crash 

This problem is fixed when I  run code ​​
`ALTER TABLE  url_alias  CHANGE `language` `langcode` VARCHAR(12) NOT NULL DEFAULT '';
`
## Steps To Reproduce

1. Setup Backdrop using Pantheon upstream https://dashboard.pantheon.io/sites/create?upstream_id=bf703821-4c18-45a1-88b8-3d9ec302273d  
2. Install Backup and migrate module 
3. Restore Drupal 7 db from /admin/config/system/backup_migrate/restore
4. Run /core/update.php

To reproduce the behavior follow steps described  in detail in this blog post https://fibonacciwebstudio.com/nota-bene/d7-backdropcms-migration-pantheon-webuisftp-option

## Actual behavior

D7 upgrade crashes if I do not modify url_alias table 

## Expected behavior

Expected behaviour is completion of the upgrade without interruption of additional steps to fix url_alias table.

**Comment by irinaz** [comment]
@indigoxela, thank you, it is good to know that code is already there!  I had this same issue very consistently on each site that I was converting on Pantheon platform.  What would be best way to troubleshoot why it is happening? This is very naughty small bug and it would be great to fix it.

**Comment by irinaz** [comment]
PHP 7.2, documentation says "The default database version for new sites is MariaDB 10.4", I can check if that is the case. 
I created fresh site, got the same error, run same db `ALTER TABLE url_alias CHANGE language langcodeVARCHAR(12) NOT NULL DEFAULT '';`
After that upgrade completed and I found the original error message(s) in the log 


` PDOException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'langcode' in 'where clause': SELECT source FROM {url_alias} WHERE alias = :alias AND langcode IN (:langcode, :language_none) ORDER BY langcode ASC, pid DESC; Array ( [:alias] => home [:langcode] => en [:language_none] => und ) in backdrop_lookup_path() (line 160 of /code/core/includes/path.inc).`
 
`EntityStorageException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'auto' in 'field list' in EntityDatabaseStorageController->save() (line 786 of /code/core/modules/entity/entity.controller.inc).`

I am not sure if this helps to identify where code related to url_alias path in  includes/update.inc , line 149 is not being executed.  There is a pointer to line 160 of /code/core/includes/path.inc, but I do not know that is helpful.
  
Thanks you all for helping with this!!

**Comment by indigoxela** [comment]
> Column not found: 1054 Unknown column 'auto' in 'field list'...

That one has its own issue: #5497 - it might also play a role here. @irinaz are you able to test the other patch to see if you get better progress without manual fixes?

**Comment by irinaz** [comment]
I will check this issue, thanks!  I also have an issue with new role "administrator" created. Will post updates later this week when I have time to do good testing

**Comment by irinaz** [comment]
@indigoxela, I will reach out to you on zulip - I am not sure what is best way to apply both patches in the same way.  thanks in advance!

**Comment by irinaz** [comment]
@indigoxela , Happy new year!  I am back on this issue hoping to figure out how it can be fixed.  I am working with the most recent version of  Backdrop CMS 1.23.1 on Pantheon and assume that patches made it to the new release.   I import database using Backup and migrate module, but still get the same error 

` PDOException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'langcode' in 'where clause': SELECT source FROM {url_alias} WHERE alias = :alias AND langcode IN (:langcode, :language_none) ORDER BY langcode ASC, pid DESC; Array ( [:alias] => home [:langcode] => en [:language_none] => und ) in backdrop_lookup_path() (line 160 of /code/core/includes/path.inc).`

Thank you very much in advance for looking into this!!


I am working with the most recent version of  Backdrop CMS 1.23.1
Install profile (pantheon)
MySQL, MariaDB, or equivalent version 5.5.30

> PHP 7.2, documentation says "The default database version for new sites is MariaDB 10.4", I can check if that is the case. I created fresh site, got the same error, run same db `ALTER TABLE url_alias CHANGE language langcodeVARCHAR(12) NOT NULL DEFAULT '';` After that upgrade completed and I found the original error message(s) in the log
> 
> ` PDOException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'langcode' in 'where clause': SELECT source FROM {url_alias} WHERE alias = :alias AND langcode IN (:langcode, :language_none) ORDER BY langcode ASC, pid DESC; Array ( [:alias] => home [:langcode] => en [:language_none] => und ) in backdrop_lookup_path() (line 160 of /code/core/includes/path.inc).`
> 
> `EntityStorageException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'auto' in 'field list' in EntityDatabaseStorageController->save() (line 786 of /code/core/modules/entity/entity.controller.inc).`
> 
> I am not sure if this helps to identify where code related to url_alias path in includes/update.inc , line 149 is not being executed. There is a pointer to line 160 of /code/core/includes/path.inc, but I do not know that is helpful.
> 
> Thanks you all for helping with this!!

**Comment by klonos** [comment]
@jenlampton I believe that you work with Pantheon, right? Have you seen this before, or can you help reproduce/troubleshoot? (I'm assuming that the D7 db needs to be from a multi-lingual site)

Wondering is the `if (db_table_exists('url_alias') && db_field_exists('url_alias', 'language'))` conditional in `update_prepare_bootstrap()` is getting itself in some race condition where the table or the filed are not loaded/detected properly 🤔

**Comment by argiepiano** [comment]
@irinaz, to me, the "red flag" in your last comment is this:

>  I import database using Backup and migrate module

If you are using the module `Backup and migrate' to import a D7 database into an already existing installation of Backdrop, you will run into all sort of issues. The problem here is that you need the upgrade process to run before any bootstrapping happens in the site. The upgrade process will take care of dropping the column `language` and adding `langcode` in `url_alias` among many other things. This happens in `update_prepare_bootstrap()`.

So, first I would do is avoid using `Backup and migrate` to import a D7 database to an existing B site. Instead, follow the tried-and-true steps. 

Now, reading the first post, it seems like you did originally follow the official steps, but still run into this issue? I don't have experience with the way Pantheon handles database backups etc, so I can't really reproduce the original issue.

**Comment by irinaz** [comment]
@argiepiano , I am trying to set up process for D7 to Backdrop Update that will work well for sitebuiders implying that almost everything can be done through UI.   I import D7 into fresh install of Backdrop either through "import" option in Pantheon UI or through Backup and Update now.  Hopefully similar work can be done using https://github.com/backdrop-contrib/d2b_migrate when issues are solved there. 

Also there is a forum post about same problem from 2019 referenced in issue summary, not sure if that is on Pantheon or not.

**Comment by argiepiano** [comment]
> I import D7 into fresh install of Backdrop either through "import" option in Pantheon UI or through Backup and Update now

You probably mean `Backup and migrate`. Using that module will never work as things are in Backdrop. The problem is that after you import your DB from the UI, Backdrop bootstraps the site without having done any of the required update process.  Plus there are tons of other issues, since you are importing a D7 database into an existing instance of a Backdrop database.

There is a reason why the [steps outlined here](https://fibonacciwebstudio.com/nota-bene/d7-backdropcms-migration-pantheon-webuisftp-option) specifically say that you have to "wipe out" all the Backdrop files and database before importing the D7 database.

**Comment by irinaz** [comment]
This is a major blocker for D7 Soft landing initiative - I would like to advocate for solution of this challenge - @quicksketch @indigoxela @klonos

**Comment by jenlampton** [comment]
> @jenlampton I believe that you work with Pantheon, right? Have you seen this before, or can you help reproduce/troubleshoot? 

I don't actually do any development on the pantheon platform, I do it locally and push my completed project up to Pantheon. I also follow the [recommended steps](https://fibonacciwebstudio.com/nota-bene/d7-backdropcms-migration-pantheon-webuisftp-option) and haven't ever tried to install a Drupal 7 site into an existing Backdrop site.

**Comment by irinaz** [comment]
@jenlampton @quicksketch , I tested importing db on local instance and on Pantheon, and it seems to be a Pantheon issue.  I will try to find a different host and import db via backup and migrate.   However, it would be huge benefit to figure out why core/update.php runs differently on Pantheon, as it is currently the hosting company that provides my favorite "one-click" install for Backdrop in the easiest possible way.

**Comment by klonos** [comment]
@irinaz I just @-mentioned you on another thread in Zulip, relevant to an issue with slow config file read/write performance, which also only seems to be an issue in Pantheon: https://backdrop.zulipchat.com/#narrow/stream/218635-Backdrop/topic/Config.20and.20distributed.20filesystem.20performance

**Comment by argiepiano** [comment]
The title of this issue is misleading - it's a symptom of a bigger problem. The main issue here is that you are trying to migrate a D7 site by importing the database (with Backup and migrate) from within an installed Backdrop site. That will never work with the current architecture. As @jenlampton mentioned in the recommended steps, the idea is that you have to completely wipe out your existing Backdrop database and config files, then import your database (Backup and migrate will not help here, since your Backdrop site is now wiped out and unbootable), and then visit /core/update.php.

Currently that's the only way in which you can upgrade a D7 site to Backdrop. The "one click" idea sounds great, but it just doesn't work with the way things are with Backdrop at the moment. Enabling the ability to import a whole D7 site into a an existing Backdrop site would take a lot of work and coding.

**Comment by irinaz** [comment]
@argiepiano , I was able to successfully run this process on my local MAMP installation. I believe that something is different in how Pantheon handles db import, maybe additional tables or some caching?

**Comment by klonos** [comment]
@argiepiano although I agree with what you said re it not being possible as it is currently, and re it requiring lots of work to make it happen, I have always found the way the D7 upgrades are being done with https://github.com/backdrop-contrib/d2b_migrate more appealing and straight-forward. But I'm digressing...

In this specific issue here, we know what the problem is (the specific db column does not have the expected name), and we have a fix that is confirmed to work. So we could explore the following options:
- at least for the 1.x cycle, anticipate the different column name, and accept both (fall back to `language` if `langcode` doesn't exist)
- if the `language` column exists, either automatically and transparently rename it, or throw a message in the status page and allow a manual trigger of the conversion (still via the UI with prompts)

I believe that we should be doing that if possible, instead of allowing the site to be "blowing up".

**Comment by klonos** [comment]
Here's an idea: I have been doing some research on the internet by searching with terms like "symlink mysql table", which initially brought different results than what my goal was (threads discussing where the db tables are being saved on disk), but then I have discovered that:

> MySQL supports views, including updatable views. Views are stored queries that when invoked produce a result set. A view acts as a virtual table.

...and also:

> Materialized views are disk based and are updated periodically based upon the query definition. Views are virtual only and run the query definition each time they are accessed.

Source: https://stackoverflow.com/questions/47098749/mysql-create-symbolic-link-to-a-table-from-another-database

Then also this: https://www.w3schools.com/mysql/mysql_view.asp

> In SQL, a view is a virtual table based on the result-set of an SQL statement.
>
> A view contains rows and columns, just like a real table. The fields in a view are fields from one or more real tables in the database.
>
> You can add SQL statements and functions to a view and present the data as if the data were coming from one single table.
>
> A view is created with the `CREATE VIEW` statement.

...which has this example:
```
CREATE VIEW view_name AS
SELECT column1, column2, ...
FROM table_name
WHERE condition;
```

So I was wondering if we could do something like this and save the day (untested - someone with better SQL knowledge should verify this):
```
CREATE VIEW language AS
SELECT langcode
FROM url_alias;
```

Which means that each time that a query would try to access the `language` table (which may not exist), we will be returning the contents of the `langcode` table, which does exist. ...or the other way around? ...or something like that anyway?

**Comment by klonos** [comment]
...close but no 😞 ...I thought that SQL views would act as "virtual columns" but they are more like "virtual tables" 🤔

**Comment by argiepiano** [comment]
@klonos, I think it's great you are giving this a try. The `d2b_migrate` module uses a very different approach: it connects to a **separate** D7 database that's separate from the site's backdrop database. That approach is workable.

The approach tried above (importing a D7 db into an existing Backdrop's site's db) is, in my view, a very complex endeavor. It may work with simple sites (perhaps that's why @irinaz was successful in her local to make this work), but I don't think it will work on a more complex site.

I bet the missing column crash is just the tip of the iceberg here. In my experience after working with upgrades when developing for Rules and Organic Groups, there will be a score or other thing that will need fixing.  Sorry to be a party pooper...

**Comment by klonos** [comment]
...apparently there is a way to create "virtual columns": https://stackoverflow.com/questions/17948940/how-to-build-virtual-columns

But if I'm reading/understanding things right, then we'd need to be querying a different table (a view) 🤔

**Comment by klonos** [comment]
> Sorry to be a party pooper...

Not a party pooper at all @argiepiano 😅  ...I don't mind exploring ideas that might make what people expect to work to actually work.

**Comment by irinaz** [comment]
@argiepiano I am using exactly same approach as d2b_migrate.  d2b_migrate is based on backup_and_migrate - I am using a **separate** D7 db (local copy or accessing through mysql) and import it via UI.

**Comment by klonos** [comment]
...last thing and I'll leave it alone for now: https://en.wikipedia.org/wiki/Virtual_column

> In relational [databases](https://en.wikipedia.org/wiki/Database) a virtual column is a table [column](https://en.wikipedia.org/wiki/Column_(database)) whose value is automatically computed using other columns values, or another deterministic expression. ...

Good news:
> MariaDB is a MySQL fork. Virtual columns were added in the 5.2 tree.

...(potentially) bad news:
> Support for virtual columns, known in MySQL as generated columns, started becoming available in MySQL 5.7. Various limitations on their use have been relaxed in subsequent versions.

**Comment by klonos** [comment]
...I've moved all my comments re virtual SQL tables/columns to #5951.

**Comment by argiepiano** [comment]
> @argiepiano I am using exactly same approach as d2b_migrate. d2b_migrate is based on backup_and_migrate - I am using a **separate** D7 db (local copy or accessing through mysql) and import it via UI.

It's unclear to me what you mean. Can you explain? So, in your existing Backdrop site, you are going to `admin/config/system/backup_migrate/restore` and restoring your D7 backup into your "Deafult Database"? That will overwrite the existing Backdrop database.   

I'm interested. Can you please explain how you are using Backup and migrate to move a D7 database into a **separate** database that's not your site's main Backdrop database?

**Comment by irinaz** [comment]
> It's unclear to me what you mean. Can you explain? So, in your existing Backdrop site, you are going to `admin/config/system/backup_migrate/restore` and restoring your D7 backup into your "Default Database"? That will overwrite the existing Backdrop database.

@argiepiano  Yes, that is what I am doing.  It is my understanding that it does overwrite the existing Backdrop database

**Comment by irinaz** [comment]
This might be related to 
https://docs.backdropcms.org/api/backdrop/core%21includes%21update.inc/function/update_prepare_bootstrap/1

**Comment by klonos** [comment]
FTR, during the dev meeting today, we said that we should start by running `select schema_version from system where name = “system“;` before and after the problem occurs, in an effort to try to pinpoint what's happening.

We also summarized the following points:
- This issue is currently only reproducible in Pantheon - not on people's locals or in any other hosting.
- It doesn't seem to be some PHP 8.x incompatibility/breakage (since @irinaz mentions that this happens in PHP 7.2).
- This happens regardless of the method to import the D7 db (either via the UI provided by Pantheon, or via the Backup and Migrate module)

**Comment by jenlampton** [comment]
> we should start by running

Some backstory: The query that Irina is running manually is supposed to be run during the upgrade process, in `update_prepare_bootstrap()`, on Pantheon, for some reason, it is not. 

That query will only run if Backdrop has determined that the database belongs to a Drupal 7 site. it does that by checking the schema version of the `system` module, which lives in in the `system` table of the database.  

The mysql query `select schema_version from system where name = “system“;` will tell us if the problem is with the value that is actually in the database at the time the code is running, or perhaps it is in one of Pantheon's fancy caching layers.  (This is my leading theory, mostly because the problem only happens on Pantheon)

Before importing the Drupal 7 database, we would expect the result to be in the 1000s. 
After it should be in the 7000s. 
After running `update.php`, it will return to the 1000s.

**Comment by irinaz** [comment]
Step 1 - Database on existing BD site -  `SELECT  `schema_version` FROM `system` WHERE `name` = "system"; ` returns 1085
Step 2 - Import D7 DB using backup and migrate - 7085

Got error 
`SQLSTATE[42S22]: Column not found: 1054 Unknown column 'role' in 'field list': SELECT role FROM {users_roles} WHERE uid = :uid; Array ( [:uid] => 1 )`
 
Try to run  /core/update.php
Got error  
`PDOException: SQLSTATE[HY000]: General error: 1034 Incorrect key file for table 'url_alias'; try to repair it: ALTER TABLE {url_alias} CHANGE `language` `langcode` VARCHAR(12) NOT NULL DEFAULT '' `

**Comment by klonos** [comment]
OK, I have started a trial account on Pantheon and have done the following:
- created and installed a vanilla D7 site to use as source
- created and installed a vanilla Backdrop site to use as destination
- inserted a bunch of silly `print()` statements in key spots in `update_prepare_bootstrap()` and pushed the code.
- enabled `update_free_access` in my settings.php and pushed the changes to Pantheon
- wiped the db and files from the Backdrop site
- created a backup of the D7 db from the Pantheon Dashboard UI
- imported the D7 db into the Backdrop site

At this point, I was able to reproduce the problem and my `print()` statements confirmed the following:
- `$system_schema` is ?? 👍🏼 
- The code within `if ($system_schema > 7000 || $system_schema < 6000) { ... }` does run 👍🏼 
- `$has_required_schema` evaluates to `1` (`TRUE` ) 👍🏼 
- The code within `if ($has_required_schema) { ... }` is executed 👍🏼 
- The code within `if (db_table_exists('url_alias') && db_field_exists('url_alias', 'language')) { ... }` is executed 👍🏼 
- I have then also added the following code to double-check things:
  ```php
    if (db_table_exists('url_alias') && db_field_exists('url_alias', 'language')) {
      print('D7 db table and columns still here');
    }
    if (db_table_exists('url_alias') && db_field_exists('url_alias', 'lancode')) {
      print('db table and columns updated to lancode');
    }
  ```
  This bit of code never got to run 🤔 

So then I looked carefully to what the error is saying, and focused on the following things specifically:
- > General error: 1034 Incorrect key file for table 'url_alias'; try to repair it ...
- > ...in db_change_field() (line 3133 of /code/core/includes/database/database.inc).

Which revealed that `db_change_field('url_alias', 'language', 'langcode', $langcode_spec, $langcode_indexes);` never actually gets to run ...I'm assuming because something "broke" the `url_alias` table (either one or both of the `db_drop_index()` calls most likely) 🤔 🤔 

Some googling of the error suggested running `REPAIR TABLE url_alias;` which should fix the issue. I tried that, but I got this error:

```
mysql> REPAIR TABLE url_alias;
+--------------------+--------+----------+---------------------------------------------------------+
| Table              | Op     | Msg_type | Msg_text                                                |
+--------------------+--------+----------+---------------------------------------------------------+
| pantheon.url_alias | repair | note     | The storage engine for the table doesn't support repair |
+--------------------+--------+----------+---------------------------------------------------------+
1 row in set (0.06 sec)
```

So I then did this:
```
mysql> ALTER TABLE url_alias ENGINE=MyISAM;
Query OK, 0 rows affected (0.07 sec)
Records: 0  Duplicates: 0  Warnings: 0
```

...and then tried again:
```
mysql> REPAIR TABLE url_alias;
+--------------------+--------+----------+----------+
| Table              | Op     | Msg_type | Msg_text |
+--------------------+--------+----------+----------+
| pantheon.url_alias | repair | status   | OK       |
+--------------------+--------+----------+----------+
1 row in set (0.06 sec)
```

Refreshed the `core/update.php` page with the error and voila! 🎉 

So, I don't know why this only seems to be happening on Pantheon, but it seems that:
- perhaps the db engine is the culprit (?)
- perhaps we need to always run `REPAIR TABLE` either at the end of `db_drop_index()` or at the beginning of `db_change_field()` (?) ...but that required the db storage engine to be MyISAM. What does it mean if we change it? Should we be changing it?

I am no expert and don't really know what would be the appropriate solution here, but I've done the legwork, so I hope that this helps move this forward.

**Comment by irinaz** [comment]
@klonos , thank you so much!  This is 100% consistent with what I see.

**Comment by klonos** [comment]
I have tried so many things/scenarios and still cannot figure out what's going on 🤔 

It certainly chokes at `db_change_field('url_alias', 'language', 'langcode', $langcode_spec, $langcode_keys_new);`, at which point the `'url_alias'` table appears to be "broken" (it has an "incorrect key file" as the message says). I have confirmed that `REPAIR`ing the table (after switching the db engine for the `'url_alias'` table to MyISAM) fixes this, but why does the table end up being "broken" in the first place? The only thing that we have done to it at this point is that we run the 2 `db_drop_index()`s 🤔 

I have tried `db_drop_unique_key('url_alias', 'pid');` and/or `db_drop_primary_key('url_alias');` and then re-adding it later, but that ended up with all sorts of other errors (like https://www.drupal.org/node/190027, which can be worded around by adding a temporary index to the table), but that didn't solve the problem either.

**Comment by klonos** [comment]
I FIXED IT!!! 🎉  ...doing a final run on a vanilla D7 -> Backdrop on Pantheon and will file a PR shortly after 🙂

**Comment by klonos** [comment]
OK, PR up for review here: https://github.com/backdrop/backdrop/pull/4325

It seems that trying to add indices in `db_change_field()` based on the same column you are trying to rename with this function causes it to choke. The solution was simple after all: rename the field without defining any indices in `db_change_field()`, then add the 2 indices separately with `db_add_index()` after `db_change_field()` has renamed the column in the database 😉

**Comment by irinaz** [comment]
@klonos , this is SO AWESOME!   
I applied patch, pushed updated core to Pantheon and used Backup_Migrate/Restore to import D7 database.  After import completed I run core/update.php, successfully updated content, user, views, etc. to Backdrop and now I have working site!! 

I got this notice during update, but update was completed.  Thank you so much for getting to the core of this issue!!!!

```
Update #1021
Failed: EntityStorageException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'langcode' in 'field list' in EntityDatabaseStorageController->save() (line 789 of /code/core/modules/entity/entity.controller.inc).
```

**Comment by klonos** [comment]
Thanks for taking the time to test and confirm @irinaz 🙏🏼 

I didn't get that error during my tests (but I was testing with a vanilla D7 db). Perhaps your D7 site had errors enabled, which carried on to the upgraded Backdrop site, which in turn allowed that error to be shown in your case. Let me test again and see.

In the meantime, can you please provide more information about your setup?
- What was your D7 db like? Was it vanilla like in my case, or did it have contrib modules, and if so, then can you please provide a list?
- What php version were you testing on? (I was testing on 7.4.30)

Noting that I had also applied @indigoxela's fix for #2518 in my test Backdrop site (although not sure if that plays any role).

**Comment by klonos** [comment]
> ... Perhaps your D7 site had errors enabled, which carried on to the upgraded Backdrop site, which in turn allowed that error to be shown in your case. ...

Turns out that my D7 source site also had all errors enabled, so that's not it. Looking further 👀

**Comment by klonos** [comment]
...nope, I cannot reproduce it. There are 3 db updates that have the number `1024`:
- `system` module: 1021 - Removes the js_cache_files variable.
- `node` module: 1021 - Creates the table to enable caching of Nodes entities.
- `user` module: 1021 - Make sure that generation of automatic path alias is enabled for user 1.

The one in system.install about the js_cache_files variable seems unrelated to the error you got, but the other 2 in node.install and user.install seem like possible candidates. I'll look into those next.

One thought that I had is that my D7 db was "too" vanilla, since D7 doesn't create any content OOTB by default. Perhaps I need to install devel_generate and generate some dummy content + some aliases and test again using that db.

**Comment by irinaz** [comment]
@klonos , it this change w is very possible that my D7 has errors in it's own configuration - site was built in 2015 and reconfigured several times :).  The most important thing is that core/update.php completes its work even when there are errors.  I hope that your fix can be committed soon!!

**Comment by klonos** [comment]
No worries @irinaz but out of curiosity, could you please:
1. run `drush pml --status=enabled` and provide the output of that command?
2. search your codebase for the string `'field list'` and repost back the files/modules where is is found? ...I was able to find only a single instance of it in Backdrop core (in `/core/modules/simpletest/tests/module.test`), and even that was in an inline comment - not actual code:

```php
    // This should not cause EXCEPTION: SQLSTATE[42S22]: Column not found: 1054
    // Unknown column 'comment_close_override' in 'field list'.
```

**Comment by quicksketch** [comment]
Great debugging @klonos! Thanks so much for finding the root here. I tried to recreate the exact SQL that gets executed when renaming the `language` column and it's a little tricky to extract with PDO doing some steps for us. I think I have it correct with the following:

```
ALTER TABLE url_alias DROP INDEX `alias_langcode_pid`;
ALTER TABLE url_alias DROP INDEX `source_langcode_pid`;

ALTER TABLE url_alias CHANGE `language` `langcode` VARCHAR(12) NOT NULL DEFAULT '' COMMENT 'The language code this alias is for; if ’und’, the alias will be used for unknown languages. Each Backdrop path can have an alias for each supported language.', ADD INDEX `alias_langcode_pid` (`alias`, `langcode`, `pid`), ADD INDEX `source_langcode_pid` (`source`, `langcode`, `pid`);
```

I tried checking if this SQL can be run on Pantheon, and it can. So it doesn't seem to be a MySQL/MariaDB issue. (General FYI: Pantheon uses MariaDB 10.0.23 by default for Backdrop sites.)

I also tried spinning up a brand [new Backdrop site on Pantheon via the upstream](https://dashboard.pantheon.io/sites/create?upstream_id=bf703821-4c18-45a1-88b8-3d9ec302273d), switching to SFTP mode, installing Devel module via project installer, and then running this, which simply pretends to alter the table by renaming "langcode" to "langcode" (everything ends up being identical):

```
// Change language column to langcode in url_alias.
db_drop_index('url_alias', 'alias_langcode_pid');
db_drop_index('url_alias', 'source_langcode_pid');
$langcode_spec = array(
  'description' => "The language code this alias is for; if 'und', the alias will be used for unknown languages. Each Backdrop path can have an alias for each supported language.",
  'type' => 'varchar',
  'length' => 12,
  'not null' => TRUE,
  'default' => '',
);
$langcode_indexes = array('indexes' =>
  array(
    'alias_langcode_pid' => array('alias', 'langcode', 'pid'),
    'source_langcode_pid' => array('source', 'langcode', 'pid'),
  ),
);
db_change_field('url_alias', 'langcode', 'langcode', $langcode_spec, $langcode_indexes);
```

This also ran without any issue.

The summary of my exploration: I can't tell why this problem is occurring :frowning_face:

As far as I can tell, the table column renaming works fine both directly in MariaDB and through PHP.

While @klonos's PR looks pretty harmless to me, I'd like to figure out if we have a larger problem with renaming table columns, or if this is something specific to the D7 site @irinaz is working on.

@irinaz Could you connect to your D7 MySQL server (prior to any upgrades) and simply run this to check the current state of the table? Maybe there's something unexpected:

```
DESCRIBE url_alias;
SHOW INDEXES FROM url_alias;
```

**Comment by klonos** [comment]
> ...or if this is something specific to the D7 site @irinaz is working on.

@quicksketch if you check my troubleshooting steps, I was able to reproduce the issue with a completely blank, vanilla, fresh  D7 installation on Pantheon. So this is nothing to do with @irinaz's D7 site, rather with whatever Pantheon is doing with its D7 installation.

> I'd like to figure out if we have a larger problem with renaming table columns...

I think that you might be able to reproduce this if you do the following:
1. create any table with a few testing columns
2. add one or more indices based on one of the columns
3. remove the indices
4. try to rename the column which previously had an index, while at the same time attempting to create a new index based on the to-be-renamed column (do all that in a single `db_change_field()` call) -> should blow up 👎🏼 
5. renaming the column first (with a `db_change_field()` call that doesn't touch any indices), and THEN adding the index **SEPARATELY** with a `db_drop_index()` should work fine 👍🏼 

Simply "pretending" to rename a column won't reproduce the issue - you need to try to actually rename a column AND set an index based on the new name of the column **AT THE SAME TIME** (all within a single `db_change_field()` call).

**Comment by irinaz** [comment]
@quicksketch , here is what I see on original db 

<img width="1207" alt="Screen Shot 2023-02-05 at 11 06 12 AM" src="https://user-images.githubusercontent.com/2497005/216839610-a56710f3-583a-43e0-97e1-1b5ed9d9e628.png">

Seems like there are some duplicate entries? 
I will do more testing with other D7 sites that we have tomorrow and report results.

**Comment by quicksketch** [comment]
Thanks @klonos, I'll try reproducing with both a column rename and new keys at the same time next.

@irinaz Thanks for the MariaDB output. That all looks correct and as expected.

**Comment by quicksketch** [comment]
Okay, confirmed the problem on Pantheon. I can reproduce it pretty directly with the following steps:

* Fresh install of Backdrop on Pantheon with the Backdrop upstream.
* Connect to MariaDB via the command line.
* Run the following to emulate a column name change:
```
CREATE TABLE url_alias2 LIKE url_alias;

ALTER TABLE url_alias2 DROP INDEX `alias_langcode_pid`;
ALTER TABLE url_alias2 DROP INDEX `source_langcode_pid`;

ALTER TABLE url_alias2 CHANGE `langcode` `langcode2` VARCHAR(12) NOT NULL DEFAULT '', ADD INDEX `alias_langcode_pid` (`alias`, `langcode2`, `pid`), ADD INDEX `source_langcode_pid` (`source`, `langcode2`, `pid`);
```

Which returns:

```
ERROR 1034 (HY000): Incorrect key file for table 'url_alias2'; try to repair it
```

I tried reproducing this on ddev with MariaDB 10, but uses `10.0.38` as opposed to `10.0.23` from Pantheon. But the same steps work fine on that version. Any newer MariaDB versions also seem to work fine. I bet anything that if we simply updated the MySQL/MariaDB version in the Pantheon upstream, this problem would "go away". The `10.0.23` version of MariaDB was released in 2015 (around the time of Backdrop 1.0)

https://pantheon.io/docs/pantheon-yml#specify-a-version-of-mariadb

I filed an issue in the Pantheon upstream issue queue at https://github.com/backdrop-ops/backdrop-pantheon/issues/68.

**Comment by klonos** [comment]
Nice! 👍🏼 

In the meantime @quicksketch, regardless of whether Pantheon updates their MariaDB version promptly or not, do you think that in order of being on the safe side of things we should simplify the code in `update_prepare_bootstrap()` to be executing the column rename and the adding of the indices separately, like I've done in my PR? (since there may be other hosting providers out there that still have outdated versions of MariaDB)

**Comment by irinaz** [comment]
Thank you SOOOO much, @quicksketch  and @klonos for doing this very deep dive - I hope that update of upsteam on Pantheon will fix this issue for everyone!!!

**Comment by klonos** [comment]
@irinaz https://github.com/backdrop-ops/backdrop-pantheon/issues/68 has been merged, so from now on Backdrop instances in Pantheon will be span up with a newer version of MariaDB, which doesn't cause this issue 😉 

@backdrop/core-committers although the specific issue with our Pantheon instances has been sorted, the fact remains that the code currently in place may break with specific versions of MariaDB (on people's local or hosting providers other than Pantheon). I believe that the code change in https://github.com/backdrop/backdrop/pull/4325 is simple and straight-forward, so it should be merged. It basically breaks down a series of tasks that were previously done in one go (rename the table and set indices based on the to-be-renamed table in a single function/query) into 2 "stages": first rename the table, then recreate the indices.

**Comment by BWPanda** [comment]
A few suggestions for that PR @klonos.

**Comment by jenlampton** [comment]
This code change looks trivial, and I think it would be a good idea in addition to the Pantheon upstream MariaDB version update. I've reviewed the code, but we will also need to run an upgrade with the patch to test.

**Comment by klonos** [comment]
@jenlampton I have tested this on Pantheon multiple times (before the upstream MariaDB version was upped), using the scenario @irinaz has been using to trigger the issue. But I guess that that doesn't count, since I was the one that provided the PR. The irony here is that now that we've updated the Backdrop recipe for Pantheon, the issue won't be easily reproducible 😅 

If people want to test this on their local with lando, know that you can provision the database with a specific version, and both MariaDB and MySQL support that in lando:
- MySQL: https://docs.lando.dev/mysql/#patch-versions (it will need to be 5.5.30 - _I think_)
- MariaDB: https://docs.lando.dev/mariadb/#patch-versions (it will need to be 10.0.23, same as the old version used in Pantheon)

**Comment by klonos** [comment]
This was discussed during the weekly dev meeting. Next steps/actions:
- confirm that this also applies to outdated MySQL versions (5.5.30)
- check other instances of `db_change_field()` where we are renaming tables and adding indices at the same time, and include the same fix in the same PR (language module seems to be doing the same)
- add a @todo item to perhaps remove/revert the fix once the version on MySQL/MariaDB is officially bumped in Backdrop (at which point a requirement notice will be shown when attempting to use these versions anyway)
- add a separate issue, to consider removing the option to change indices from `db_change_field()` (although that would be a BC breaking change, so slated for 2.x)

**Comment by klonos** [comment]
> ...we will also need to run an upgrade with the patch to test. (removing RTBC label until that's done, just so it's not committed accidentally)

@jenlampton noting that @quicksketch mentioned that he has also tested this with DDEV, and have found that a) the PR fixes the problem and does not cause any new problems (see https://youtu.be/t0o9lxSsTmc?t=2245)

**Comment by laryn** [comment]
I can confirm I ran into this while upgrading locally with DDEV and this patch fixed it.

**Comment by laryn** [comment]
@irinaz I am coming late to the issue but it looks like there's a list of next steps here:
https://github.com/backdrop/backdrop-issues/issues/5496#issuecomment-1452664030

I'm not certain which (if any) need to happen before this is ready to go. @klonos do you have time to chime in here quickly (no worries if not).

**Comment by klonos** [comment]
Sorry for failing to reply here earlier @laryn 😞 

Unfortunately, I cannot help with this step/action here:
> - confirm that this also applies to outdated MySQL versions (5.5.30)

I have tried setting up my lando instance with that specific version of MySQL, but it is not supported/provided any more. That means that it would require a considerable amount of time and manual effort on my end to set up a local with that version. If anyone else is able to do that, then please do so when you have time/capacity. (steps to reproduce this are in the issue summary - and thank you in advance 🙏🏼 ). Point being that I don't think that we should be holding this fix back on account of a version that is outdated, when we have tested and confirmed that the proposed PR fixes the problem for those coming across it, while at the same time it is not breaking anything (the workaround simply breaks down a single task that did two things at the same time into those two things individually).

In the meantime, I have:
- rebased the PR
- implemented the same workaround to other instances of `db_change_field()` that were adding indexes at the same time (as previously requested by @quicksketch)
- added `@todo` items to consider reverting the workaround once the minimum version of MySQL/MariaDB in the Backdrop official system requirements is bumped to a version that doesn't exhibit this problem.

Ready for review/consideration.

Since with this PR we will now be implementing this workaround in a few places, it feels that perhaps it should be implemented in an internal function. That would be easier to remove later on 🤔 ...I'll bring this up in the dev meeting today, but more generally, I'm interested to see what our @backdrop/core-committers think.

**Comment by irinaz** [comment]
Hi, 
I am running an upgrade on the most recent version of Backdrop, and I see the following error.   I am not sure if this is the same problem as we were dealing with in the past - I do not see any reference to table url_alias, and I see new reference to 'rid'.   

Do you know if I should open new issue, or this problem might have the same root as problem that we have in description? 

`SQLSTATE[42S22]: Column not found: 1054 Unknown column 'langcode' in 'where clause': SELECT redirect.`rid` AS `rid` FROM {redirect} redirect WHERE ( (`source` LIKE :db_condition_placeholder_0 ESCAPE '\\') OR (`source` = :db_condition_placeholder_1) )AND (`langcode` IN (:db_condition_placeholder_2, :db_condition_placeholder_3)) ; Array ( [:db_condition_placeholder_0] => home [:db_condition_placeholder_1] => [:db_condition_placeholder_2] => en [:db_condition_placeholder_3] => und )`

**Comment by klonos** [comment]
...I've fixed the conflicts in the PR but I'm now getting another PDOException I need to look into. So this is still NW.

**Comment by herbdool** [comment]
I recently (this month) used this pr to get past fatal errors in an upgrade I was doing in Lando (using 10.5.19-MariaDB). I had imported the db with `lando db-import` and ran the upgrade script `bee updatedb` where it failed after awhile. (I should have kept better notes but it was launch day). I hadn't experienced it in my previous test of the upgrade so I'm not sure what was different. After applying this PR I got past the fatal error and could continue with `bee updatedb`.

**Comment by klonos** [comment]
> ... I hadn't experienced it in my previous test of the upgrade so I'm not sure what was different. ...

I'm pretty sure that it has to do with MySQL versions @herbdool. So if you can check and provide the versions being used in the environments where you face the problem vs where it went smoothly, perhaps it'd help us confirm that theory.

**Comment by jenlampton** [comment]
The only way this error should happen is if the module doing the update is doing it at the wrong time. 

@irinaz  can you check your module for an implementation of `https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21system.api.php/function/hook_update_dependencies/1`?

 It needs to use that hook to put its own update before the core update where `language` gets converted to `langcode`.

Is it possible that different versions of mysql are operating at different speeds , and in some cases these"just happen" to be in the right order, but in others they don't?

**Comment by irinaz** [comment]
@jenlampton , I am cross posting another issue related to langcode - https://github.com/backdrop-contrib/d2b_migrate/issues/45. I do not know which version of db some of sites are using when I work on an instance provided by a customer (I could find out if needed).

**Comment by jenlampton** [comment]
@irinaz There's no need to get any additional information from your sites, it's going to be the same everywhere. 

Running Backdrop CMS on a database that isn't a backdrop database isn't supported. There's no bug here, so I'm going to close the issue. 

The solution is to convert your Drupal 7 database to a Backdrop database **before** visiting any pages in Backdrop, as documented. Once the database is updated, you won't see the error. 

If you want to use the software in ways it wasn't intended you are welcome to, but be prepared for strange problems to crop up (and not just this issue, any number of other things could happen). There is no need to open core issues when you encounter problems while trying to load a Drupal 7 database into a running backdrop site. Contrib is the right place to try these unsupported workflows and work through all the potential issues that crop up :)

**Comment by klonos** [comment]
FTR: I've opened #6554 to deal with the bug with `db_change_field()` and older MariaDB versions.
